processing/src/main/java/io/druid/query/QueryToolChest.java

```diff
@@ -36,11 +46,81 @@
 {
   /**
+   * @param filter Filter input to optimize
+   *
+   * @return Returns an optimized {@link DimFilter}
+   * Optimization is done by rewriting all ExtractionDimFilter of type {@link LookupExtractionFn} as a clause of literals {@link SelectorDimFilter}
+   */
+  public static DimFilter optimizeExtractionDimFilter(final DimFilter filter)
+  {
+    if (filter instanceof NotDimFilter) {
+      return Druids.newNotDimFilterBuilder().field(optimizeExtractionDimFilter(((NotDimFilter) filter).getField()))
+                   .build();
+    } else if (filter instanceof AndDimFilter) {
+      return Druids.newAndDimFilterBuilder()
+                   .fields(Lists.transform(((AndDimFilter) filter).getFields(), new Function<DimFilter, DimFilter>()
+                   {
+                     @Nullable
+                     @Override
+                     public DimFilter apply(@Nullable DimFilter input)
+                     {
+                       return optimizeExtractionDimFilter(input);
+                     }
+                   })).build();
+    } else if (filter instanceof OrDimFilter) {
+      return Druids.newOrDimFilterBuilder()
+                   .fields(Lists.transform(((OrDimFilter) filter).getFields(), new Function<DimFilter, DimFilter>()
+                   {
+                     @Nullable
+                     @Override
+                     public DimFilter apply(@Nullable DimFilter input)
+                     {
+                       return optimizeExtractionDimFilter(input);
+                     }
+                   })).build();
+    } else if (okToOptimize(filter)) {
+      ExtractionDimFilter extractionDimFilter = (ExtractionDimFilter) filter;
+      LookupExtractor lookup = ((LookupExtractionFn) extractionDimFilter.getExtractionFn()).getLookup();
+      List<String> keys = lookup.unapply(extractionDimFilter.getValue());
+      final String dimensionName = extractionDimFilter.getDimension();
+      if (!keys.isEmpty()) {
+        return new OrDimFilter(Lists.transform(keys, new Function<String, DimFilter>()
+        {
+          @Nullable
+          @Override
+          public DimFilter apply(@Nullable String input)
+          {
+            return new SelectorDimFilter(dimensionName, input);
+          }
+        }));
+      } else {
+        // Maybe return a noop Filter
+        return filter;
+      }
+
+    }
+    return filter;
+  }
+
+  private static boolean okToOptimize(DimFilter filter){
+    ExtractionDimFilter extractionDimFilter = null;
+    LookupExtractionFn lookupExtractionFn = null;
+    if (filter instanceof ExtractionDimFilter){
+      extractionDimFilter = (ExtractionDimFilter) filter;
+      if (extractionDimFilter.getExtractionFn() instanceof  LookupExtractionFn){
+        lookupExtractionFn = (LookupExtractionFn) extractionDimFilter.getExtractionFn();
+      }
+    }
+    return extractionDimFilter != null && lookupExtractionFn != null && lookupExtractionFn.isOptimize();
```

Comment:
why lookupExtractionFn.isOptimize() is needed, it appears if unApply() is supported then this optimization is always a good idea. what is the catch?